### PR TITLE
Splitting reward contract 

### PR DESCRIPTION
**Pull request by vistra**
Into 3:
- Staking rewards
- Bootstrap rewrads
- Fees

Mssing funtionality:
- all related to kyc/general committee

### CODE REVIEW

**Comment by vistra** [comment]
@OdedWx @ronnno please review, made some changes to the interfaces

File: contracts/BootstrapRewards.sol [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    uint256 pool;
+
+    // TODO - should the fixed pool rate be a function of the committee size?
+    uint256 poolMonthlyRate; // todo - deprecate this
+    uint256 generalCommitteeMonthlyRate; // todo - apply this rate
+    uint256 complianceCommitteeMonthlyRate; // todo - apply this rate
+
+    uint256 lastPayedAt;
+
+    mapping(address => uint256) externalTokenBalance;
```
Comment (OdedWx) [inline contracts/BootstrapRewards.sol]: rename to bootStrapBalance

File: contracts/BootstrapRewards.sol [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        _assignRewards();
+    }
+
+    function _assignRewards() private {
+        // TODO we often do integer division for rate related calculation, which floors the result. Do we need to address this?
+        // TODO for an empty committee or a committee with 0 total stake the divided amounts will be locked in the contract FOREVER
+
+        uint256 duration = now.sub(lastPayedAt);
+
+        uint256 amount = Math.min(poolMonthlyRate.mul(duration).div(30 days), pool);
```
Comment (OdedWx) [inline contracts/BootstrapRewards.sol]: uses poolMonthlyRate and instead of generalCommitteeMonthlyRate

File: contracts/BootstrapRewards.sol [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        contractRegistry = _contractRegistry;
+    }
+
+    // todo - deprecate this, use setGeneralCommitteeBootstrapMonthlyRate and setComplianceCommitteeBootstrapMonthlyRate instead
+    function setPoolMonthlyRate(uint256 rate) external onlyRewardsGovernor {
+        _assignRewards();
+        poolMonthlyRate = rate;
+    }
+
+    function setGeneralCommitteeBootstrapMonthlyRate(uint256 rate) external {
```
Comment (OdedWx) [inline contracts/BootstrapRewards.sol]: Monthly is awkward (30 days?) modify to Annual

File: contracts/BootstrapRewards.sol [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+    function assignAmountFixed(uint256 amount) private {
+        address[] memory currentCommittee = _getCommittee();
+
+        uint256[] memory assignedRewards = new uint256[](currentCommittee.length);
+
+        uint256 totalAssigned = 0;
+
+        for (uint i = 0; i < currentCommittee.length; i++) {
+            uint256 curAmount = amount.div(currentCommittee.length);
```
Comment (OdedWx) [inline contracts/BootstrapRewards.sol]: the bootstrap amount is per validator, not committee

File: contracts/Fees.sol [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+import "./IStakingContract.sol";
+import "./interfaces/IElections.sol";
+import "./spec_interfaces/IFees.sol";
+
+contract Fees is IFees, Ownable {
+    using SafeMath for uint256;
+
+    enum ComplianceType {
```
Comment (OdedWx) [inline contracts/Fees.sol]: CommiteeType ?

File: contracts/BootstrapRewards.sol (hunk 2 above)
Comment (vistra) [reply contracts/BootstrapRewards.sol]: `30 days` in solidity is the number of seconds in 30 days, which is 2592000. Why 86400?

File: contracts/BootstrapRewards.sol [hunk 6]
```diff
@@ -39,6 +39,7 @@ contract BootstrapRewards is IBootstrapRewards, Ownable {
         require(address(_externalToken) != address(0), "externalToken must not be 0");
 
         externalToken = _externalToken;
+        // TODO - The initial lastPayedAt should be set in the first assignRewards.
```
Comment (vistra) [inline contracts/BootstrapRewards.sol]: Which time period will the first call to assignRewards will use?

Comment (OdedWx) [reply contracts/BootstrapRewards.sol]: none / lastPayedAt = now.
The first call will init it, as the contracts are likely to be deployed prior to enablement

File: contracts/BootstrapRewards.sol (hunk 2 above)
Comment (OdedWx) [reply contracts/BootstrapRewards.sol]: sec in a day. can ignore

File: contracts/BootstrapRewards.sol (hunk 6 above)
Comment (vistra) [reply contracts/BootstrapRewards.sol]: So maybe a separate enablement method?

Comment (vistra) [reply contracts/BootstrapRewards.sol]: (so enablement can only be done by governance and not by everyone)

**Pull request merged**